### PR TITLE
[DDK] Fix compilation error with g++ by using unnamed structs in anonymous union

### PR DESCRIPTION
**Pull request by assorted**
## Purpose

Fix the .cpp code compilation using sdk\include\ddk\scsi.h. This is necessary for the UDFS driver written in C++.

JIRA issue: N/A

### CODE REVIEW

**Comment by HBelusca** [comment]
How do you get this error? Our build bots don't show any build error regarding the problem you claim is happening :thinking:

**Review by HBelusca — CHANGES_REQUESTED** [review]
These changes would make the symbols incompatible with official MS definitions from the DDK/WDK:
https://github.com/tpn/winsdk-10/blob/master/Include/10.0.16299.0/shared/scsi.h#L502

**Comment by assorted** [comment]
I will figure out why user-mode applications compile normally, but drivers do not. It might be a matter of standards or compiler settings. Perhaps a workaround can be found.

**Comment by HBelusca** [comment]
@assorted Could you maybe also tell us about your build environment?

**Comment by assorted** [comment]
https://github.com/mingw-w64/mingw-w64/blob/48cf32beae4d400cf6a7f1d57ab51c2b409a9f44/mingw-w64-headers/ddk/include/ddk/scsi.h#L1126 For example, in this repository, a similar problem has already been fixed.
The issue arises when trying to include scsi.h in any C++ project. During compilation, we get the error "error: 'struct _PLAY_CD::::_LBA' invalid; an anonymous union may only have public non-static data members [-fpermissive]". I encountered this problem while trying to include it in my UDFS driver here. 
By the way, g++ will not be able to compile "MS definitions from the DDK/WDK". The __extension__ attribute, which is applied through NONAMELESSUNION, won't help here. g++ simply does not allow public struct declarations within a union.

**Comment by assorted** [comment]
Sorry for the trouble. #define NONAMELESSUNION helped.

**Comment by assorted** [comment]
```
#include <ntifs.h>

// HACK: We need to work around the issue of using named structures in 
// anonymous unions in the scsi.h module. Using the NONAMELESSUNION flag 
// for the entire project is a very bad idea, as it would require the 
// entire project code to use the macro _STRUCT_NAME(u.) to access the 
// fields of such unions.

#pragma push_macro("DUMMYUNIONNAME")
#undef DUMMYUNIONNAME
#define DUMMYUNIONNAME u
#include <scsi.h>
#pragma pop_macro("DUMMYUNIONNAME")
```
I used this hack to work around compilation errors.  This way, I was able to bypass two of them. However, the anonymous one at the very end does not have [DUMMYUNIONNAME ](https://github.com/reactos/reactos/blob/90de2b660b12919c2f60f2f31ad41e37c123d662/sdk/include/ddk/scsi.h#L3342)

Thus, the current scsi.h cannot be compiled in GNU C++ even in NONAMELESSUNION mode. I believe it would be correct to simply remove the public structure names from the anonymous unions. By the way, this has been done in the mingw-w64 project [mingw-w64](https://github.com/mingw-w64/mingw-w64/blob/48cf32beae4d400cf6a7f1d57ab51c2b409a9f44/mingw-w64-headers/ddk/include/ddk/scsi.h#L1126)
@HBelusca Please let me know what you think.

**Comment by HBelusca** [comment]
I would instead argue that if we have to remove or fix stuff, it would be the GCC hacks needed to make code compile (i.e. adjusting these ANONYMOUS_UNION things?), instead of the public struct names.

Do you compile your code with `-fms-extensions` ? If not, would it change some of this behaviour?

Cc @ThFabba @tkreuzer @binarymaster @AmineKhaldi for these questions pertaining to SDK/DDK headers.

**Comment by assorted** [comment]
No attributes or build options will allow this code to compile in GCC C++. The _-fms-extensions_ or any attributes won’t help because the C++ standard prohibits declaring structures within anonymous unions, and GCC strictly adheres to the standard. However, it compiles perfectly in MSVC or GCC C. By the way, the NONAMELESSUNION hack is intended for some ancient compilers that do not support anonymous structures at all, but in the ReactOS project, it is used to work around a problem similar to mine. Here are all the possible solutions:

1. Fix all the header files in the DDK, as I did for ddk\scsi.h.
2. Add the missing DUMMYUNIONNAME in the DDK and use NONAMELESSUNION to convert anonymous unions into named ones with the name "u". However, the entire code would then be messy, like this:
```
dest._STRUCT_NAME(u.)LowPart = s._STRUCT_NAME(u.)LowPart & m._STRUCT_NAME(u.)LowPart;
```
3. Acknowledge that the DDK is not compatible with GNU C++ and rewrite the single UDFS driver in pure C. This isn’t a problem, but it looks like a step backward.

**Comment by HBelusca** [comment]
> No attributes or build options will allow this code to compile in GCC C++. The _-fms-extensions_ or any attributes won’t help because the C++ standard prohibits declaring structures within anonymous unions, and GCC strictly adheres to the standard. However, it compiles perfectly in MSVC or GCC C.

Indeed, I confirm your observations. I suppose MSVC "in C++ mode" would compile that code OK as well?

> By the way, the NONAMELESSUNION hack is intended for some ancient compilers that do not support anonymous structures at all, but in the ReactOS project, it is used to work around a problem similar to mine. 

That's true.

> Here are all the possible solutions:
> 
>     1. Fix all the header files in the DDK, as I did for ddk\scsi.h.

Do you have a list of all those DDK header files that might require a similar fix?

>     2. Add the missing DUMMYUNIONNAME in the DDK and use NONAMELESSUNION to convert anonymous unions into named ones with the name "u". [...]

This could help; but using NONAMELESSUNION is actually not the general rule in our codebase as well, but the exception (for wine-imported code or so, as far as I could see), so the problems you've got could still happen.

>     3. Acknowledge that the DDK is not compatible with GNU C++

Definitively true! But that's why MinGW(-w64) and us (and maybe Wine too? ... but I guess they may have a more liberal approach than us in regards to trying to keep compatibility with public MS headers) try to add workarounds for GCC/G++.

> and rewrite the single UDFS driver in pure C. This isn’t a problem, but it looks like a step backward.

I haven't checked how much C++-only syntax they are using here, or whether it's just C code inside files with .cpp extension ^^, but otherwise this would kind of defeat the whole purpose of @Alter-1 et al. code.

----

If the list of files for point 1 is short (a couple of them), it may be of interest to "fix" them (and maybe implement part of the point 2); otherwise I don't think it would be the best approach. Also, knowing that MS is nowadays compiling parts of kernel drivers in C++ (see their WDF etc.), it _might_ happen that they fixed some of their WDK headers as well -- something to keep an eye on.

I don't have a definitive answer to give to you, I would like to hear what the SDK/DDK headers experts here have to say.

File: sdk/include/ddk/scsi.h [hunk 1]
```diff
@@ -1239,11 +1239,11 @@ typedef union _CDB {
     UCHAR ExpectedSectorType:3;
     UCHAR Lun:3;
     _ANONYMOUS_UNION union {
-      struct _LBA {
+      struct {
```
Comment (tkreuzer) [inline sdk/include/ddk/scsi.h:1242]: This is not what the native headers do and removing these tags can easily break other code that relies on them. It seems like GCC has some retardedness going on as usual, making up issues where there are none. Anyway we can work around this in a way that is compatible with the original definition:
Move these struct declarations to before the main struct declaration and replace this this with `struct _LBA LBA;`

Comment (assorted) [reply sdk/include/ddk/scsi.h:1242]: Thanks for the idea. You suggest moving the declaration of the `struct _LBA {...};` structure to the global scope. This is also not very good.

**Comment by assorted** [comment]
I couldn't find a better solution than removing the _LBA and _MSF tags. Moving the structure declaration out of the union as suggested above causes a conflict with the Reserved1 field. Closing the merge request. I will use compiler options `target_compile_options(udfs PRIVATE -Wno-error -fpermissive)`

**Pull request closed without merging**